src/engine.rs

```diff
@@ -414,6 +448,138 @@ impl Engine {
+        receiver: Address,
+        value: Wei,
+        gas_limit: u64,
+    ) -> EngineResult<SubmitResult> {
+        self.call(sender, receiver, value, Vec::new(), gas_limit)
+    }
+
+    /// Mint tokens for recipient on a particular ERC20 token
+    /// This function should return the amount of tokens unused,
+    /// which will be always all (<amount>) if there is any problem
+    /// with the input, or 0 if tokens were minted successfully.
+    ///
+    /// The output will be serialized as a String
+    /// https://github.com/near/NEPs/discussions/146
+    ///
+    /// IMPORTANT: This function should not panic, otherwise it won't
+    /// be possible to return the tokens to the sender.
+    pub fn receive_erc20_tokens(&mut self, args: NEP141FtOnTransferArgs, gas_limit: u64) {
+        let str_amount = args.amount.to_string();
+        let output_on_fail = str_amount.as_bytes();
+
+        let token = sdk::predecessor_account_id();
+
+        // Parse message to determine recipient and fee
+        let (recipient, fee) = {
+            // Message format:
+            //      Recipient of the transaction - 40 characters (Address in hex)
+            //      Fee to be paid in ETH (Optional) - 64 characters (Encoded in big endian / hex)
+            let mut message = args.msg.as_bytes();
+            assert_or_finish!(message.len() >= 40, output_on_fail);
+
+            let recipient = Address(unwrap_res_or_finish!(
+                hex::decode(&message[..40]).unwrap().as_slice().try_into(),
+                output_on_fail
+            ));
+            message = &message[40..];
+
+            let fee = if message.is_empty() {
+                U256::from(0)
+            } else {
+                assert_or_finish!(message.len() == 64, output_on_fail);
+                U256::from_big_endian(
+                    unwrap_res_or_finish!(hex::decode(message), output_on_fail).as_slice(),
+                )
+            };
+
+            (recipient, fee)
+        };
+
+        let erc20_token = Address(unwrap_res_or_finish!(
+            unwrap_res_or_finish!(
+                self.get_erc20_from_nep141(token.as_slice()).ok_or(()),
+                output_on_fail
+            )
+            .as_slice()
+            .try_into(),
+            output_on_fail
+        ));
+
+        if fee != U256::from(0) {
+            let relayer_account_id = sdk::signer_account_id();
+            let relayer_address = unwrap_res_or_finish!(
+                self.get_relayer(relayer_account_id.as_slice()).ok_or(()),
+                output_on_fail
+            );
+
+            unwrap_res_or_finish!(
+                self.transfer(
+                    recipient.clone(),
+                    relayer_address.clone(),
+                    Wei::new_u64(fee.as_u64()),
+                    gas_limit
+                ),
+                output_on_fail
+            );
+        }
+
+        let selector = &sdk::keccak("mint(address,uint256)".as_bytes())[..4];
```

Comment:
To do properly we need something like this:

```
const fn selector() -> Vec<u8> {
    let mut hasher = sha3::Keccak256::new();
    hasher.update("mint(address,uint256)".as_bytes());
    hasher.finalize()[..4].to_vec()
}
```

But this fails to compile with: 

```
error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
  --> src/engine.rs:26:22
   |
26 |     let mut hasher = sha3::Keccak256::new();
   |                      ^^^^^^^^^^^^^^^^^^^^^^

error[E0658]: mutable references are not allowed in constant functions
  --> src/engine.rs:27:5
   |
27 |     hasher.update("mint(address,uint256)".as_bytes());
   |     ^^^^^^
   |
   = note: see issue #57349 <https://github.com/rust-lang/rust/issues/57349> for more information
   = help: add `#![feature(const_mut_refs)]` to the crate attributes to enable

error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
  --> src/engine.rs:27:5
   |
27 |     hasher.update("mint(address,uint256)".as_bytes());
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
  --> src/engine.rs:28:5
   |
28 |     hasher.finalize()[..4].to_vec()
   |     ^^^^^^^^^^^^^^^^^

error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
  --> src/engine.rs:28:5
   |
28 |     hasher.finalize()[..4].to_vec()
   |     ^^^^^^^^^^^^^^^^^^^^^^

error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
  --> src/engine.rs:28:5
   |
28 |     hasher.finalize()[..4].to_vec()
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

An alternative is hardcode this value (since it will not change), add a comment about how it was created and add test that confirms it is the expected value.

Reply:
> An alternative is hardcode this value (since it will not change), add a comment about how it was created and add test that confirms it is the expected value.

Yes, this is what I meant. Sorry for the ambiguity.

Reply:
Done.